backend/packages/Upgrade/src/api/services/ExperimentAssignmentService.ts

```diff
@@ -905,42 +928,19 @@ export class ExperimentAssignmentService {
 
     if (groupCount && userCount && experiment.assignmentUnit === ASSIGNMENT_UNIT.GROUP) {
       // fetch all the monitored document if exist
-      const monitoredDocuments = await getMonitoredDocumentOfExperiment(experiment);
-
-      // check for student inside each group
-      const groupMap = new Map<string, Set<string>>();
-      // groupAssignments.forEach((groupAssignment) => {
-      //   groupMap.set(groupAssignment.groupId, []);
-      // });
-      monitoredDocuments.forEach((monitoredDocument) => {
-        const groupId = monitoredDocument.user.workingGroup[group];
-        // create new Set for new group
-        if (!groupMap.has(groupId)) {
-          groupMap.set(groupId, new Set());
-        }
-        groupMap.set(groupId, groupMap.get(groupId).add(monitoredDocument.user.id));
-      });
-      let groupSatisfied = 0;
-
-      groupMap.forEach((groupElement) => {
-        if (groupElement.size >= userCount) {
-          groupSatisfied++;
-        }
-      });
+      const monitoredDocuments = await getMonitoredDocumentOfExperimentTypeGroup(experiment);
 
-      if (groupSatisfied >= groupCount) {
+      // check if the group count is more than ending criteria
+      if (Object.keys(monitoredDocuments).length >= groupCount) {
         await this.experimentRepository.updateState(experiment.id, EXPERIMENT_STATE.ENROLLMENT_COMPLETE, undefined);
         await this.stateTimeLogsRepository.save(stateTimeLogDoc);
       }
       // check the individual assignment table for the user
     } else if (userCount) {
       // fetch all the monitored document if exist
-      const monitoredDocuments = await getMonitoredDocumentOfExperiment(experiment);
-      const userIds = monitoredDocuments.map((doc) => {
-        return doc.user.id;
-      });
-      const uniqueUser = new Set(userIds);
-      if (uniqueUser.size >= userCount) {
+      const monitoredDocuments = await getMonitoredDocumentOfExperimentTypeIndividual(experiment);
+
+      if (monitoredDocuments[0].count >= userCount) {
```

Comment:
is there a possibility that `monitoredDocuments` can be empty array? if so, this will throw syntax error, as index 0 won't have an object with property of `count`.

Reply:
if we can create a `MonitoredDocument` type and set the explicit return type of `getMonitoredDocumentOfExperimentTypeIndividual` as `MonitoredDocument[]`, it may be helpful.

Reply:
You have a fair point. Thank you. I will update the code.